src/Playwright/Transport/WebSocketTransport.cs

```diff
@@ -0,0 +1,180 @@
+            _options = options;
+            _slowMo = _options?.SlowMo != null ? (int)_options?.SlowMo : 0;
+            _timeout = _options?.Timeout != null ? (int)_options?.Timeout : 30000;
+            _readerCancellationSource = new CancellationTokenSource(_timeout);
+            SetRequestHeaders();
+            _ = ConnectAsync();
+        }
+
+        /// <inheritdoc cref="IDisposable.Dispose"/>
+        ~WebSocketTransport() => Dispose(false);
+
+        public event EventHandler<MessageReceivedEventArgs> MessageReceived;
+
+        public event EventHandler<LogReceivedEventArgs> LogReceived;
+
+        public event EventHandler<TransportClosedEventArgs> TransportClosed;
+
+        public bool IsClosed { get; private set; }
+
+        public async Task SendAsync(string message)
+        {
+            await Task.Delay(_slowMo).ConfigureAwait(false);
+
+            try
+            {
+                var messageBuffer = Encoding.UTF8.GetBytes(message);
+                if (!_readerCancellationSource.IsCancellationRequested)
+                {
+                    await _webSocket.SendAsync(new ArraySegment<byte>(messageBuffer, 0, messageBuffer.Length), WebSocketMessageType.Text, true, _readerCancellationSource.Token).ConfigureAwait(false);
+                }
+            }
+            catch (Exception ex)
+            {
+                Close(ex);
+            }
+        }
+
+        /// <inheritdoc/>
+        public void Close(string closeReason)
+        {
+            if (!IsClosed)
+            {
+                IsClosed = true;
+                HandleSocketClose(closeReason);
+            }
+        }
+
+        /// <inheritdoc/>
+        public void Dispose()
+        {
+            Dispose(true);
+            GC.SuppressFinalize(this);
+        }
+
+        private void Close(Exception ex)
+        {
+            System.Diagnostics.Debug.WriteLine(ex);
+            Close(ex.ToString());
+        }
+
+        private void HandleSocketClose(string closeReason)
+        {
+            TransportClosed?.Invoke(this, new TransportClosedEventArgs { CloseReason = closeReason });
+            _readerCancellationSource?.Cancel();
+        }
+
+        private async Task ConnectAsync()
+        {
+            await _webSocket.ConnectAsync(new Uri(_wsEndpoint), _readerCancellationSource.Token).ConfigureAwait(false);
+            ScheduleTransportTask(DispatchIncomingMessagesAsync, _readerCancellationSource.Token);
+        }
+
+        private void ScheduleTransportTask(Func<CancellationToken, Task> func, CancellationToken cancellationToken)
+            => Task.Factory.StartNew(() => func(cancellationToken), cancellationToken, TaskCreationOptions.LongRunning, TaskScheduler.Current);
+
+        private async Task DispatchIncomingMessagesAsync(CancellationToken token)
+        {
+            try
+            {
+                if (_webSocket.State == WebSocketState.Closed)
+                {
+                    Close("Closed");
```

Comment:
This is probably the third time I'm making this comment - we should not react to the fact that socket is closed with an attempt to Close the socket. This is where you call HandleSocketClosed()